common/cache/domainCache.go

```diff
@@ -199,39 +191,43 @@ func (c *domainCache) GetAllDomain() map[string]*DomainCacheEntry {
 // WARN: the beforeCallback function will be triggered by domain cache when holding the domain cache lock,
 // make sure the callback function will not call domain cache again in case of dead lock
 // afterCallback will be invoked when NOT holding the domain cache lock.
-func (c *domainCache) RegisterDomainChangeCallback(shard int, initialNotificationVersion int64, beforeCallback CallbackFn, afterCallback CallbackFn) {
+func (c *domainCache) RegisterDomainChangeCallback(shard int, initialNotificationVersion int64,
+	prepareCallback PrepareCallbackFn, callback CallbackFn) {
 	c.Lock()
-	c.beforeCallbacks[shard] = beforeCallback
-	c.afterCallbacks[shard] = afterCallback
-	domainNotificationVersion := c.domainNotificationVersion
+	c.prepareCallbacks[shard] = prepareCallback
+	c.callbacks[shard] = callback
 	c.Unlock()
 
 	// this section is trying to make the shard catch up with domain changes
-	if domainNotificationVersion > initialNotificationVersion {
-		domains := DomainCacheEntries{}
-		for _, domain := range c.GetAllDomain() {
-			domains = append(domains, domain)
-		}
-		// we mush notify the change in a ordered fashion
-		// since history shard have to update the shard info
-		// with domain change version.
-		sort.Sort(domains)
-		for _, domain := range domains {
-			if domain.notificationVersion >= initialNotificationVersion {
-				beforeCallback(nil, domain)
-				afterCallback(nil, domain)
-			}
+	domains := DomainCacheEntries{}
+	for _, domain := range c.GetAllDomain() {
+		domains = append(domains, domain)
+	}
+	// we mush notify the change in a ordered fashion
+	// since history shard have to update the shard info
+	// with domain change version.
+	sort.Sort(domains)
+
+	prevEntries := []*DomainCacheEntry{}
+	nextEntries := []*DomainCacheEntry{}
+	for _, domain := range domains {
+		if domain.notificationVersion >= initialNotificationVersion {
+			prevEntries = append(prevEntries, nil)
+			nextEntries = append(nextEntries, domain)
 		}
 	}
+	if len(prevEntries) > 0 {
+		prepareCallback()
```

Comment:
Can we simplify this have just single callback?  We seems to be calling them immediately one after another.  I don't see why we need to register 2 separate callbacks.

Reply:
this one is the catch up logic, meaning that there is no domain change involved when shard register the callback.

the actual logic is shown below, basically pause all shards, then do domain updates, then trigger a snapshot of shard ack levels from min to max and start failover.
this will guarantee that all task after the min ack level will be progressed as active task at least once.

from:
https://github.com/uber/cadence/pull/1280/files/886b3449137fd6db32f87827cb716f2a534cb503#diff-f8e74f79013d8fa5cd8f4fa32c12d21fR316
to:
https://github.com/uber/cadence/pull/1280/files/886b3449137fd6db32f87827cb716f2a534cb503#diff-f8e74f79013d8fa5cd8f4fa32c12d21fR337